File: TinyScript/java/src/main/java/parser/ast/CallExpr.java
package parser.ast;

import lexer.TokenType;
import parser.util.ParseException;
import parser.util.PeekTokenIterator;

public class CallExpr extends Expr {
    public CallExpr() {
        super();
        this.label = "call";
        this.type = ASTNodeTypes.CALL_EXPR;

    }

    public static ASTNode parse(ASTNode factor, PeekTokenIterator it) throws ParseException {
        var expr = new CallExpr();
        expr.addChild(factor);

        it.nextMatch("(");
        ASTNode p = null;
        while((p = Expr.parse(it)) != null) {
            expr.addChild(p);
            if(!it.peek().getValue().equals(")")) {
                it.nextMatch(",");
            }
        }
        it.nextMatch(")");
        return expr;
    }
}

File: TinyScript/js/src/parser/ast/CallExpr.js
const Expr = require('./Expr')
const ASTNodeTypes = require("./ASTNodeTypes")
class CallExpr extends Expr {
    constructor() {
        super()
        this.type = ASTNodeTypes.CALL_EXPR
        this.label = "call"
    }

}

module.exports = CallExpr 

CallExpr.parse = (factor, it) => {
    const expr = new CallExpr()
    expr.addChild(factor)

    it.nextMatch("(")
    let p = null
    while((p = Expr.parse(it)) != null) {
        expr.addChild(p)
        if(!it.peek().getValue() === ")") {
            it.nextMatch(",")
        }
    }
    it.nextMatch(")")
    return expr
}
